addons/currency_rate_live/models/yahoo_finance.py

```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8 -*-
+import json
+import logging
+import urllib
+
+from openerp import _, api, models
+from openerp import exceptions
+
+_logger = logging.getLogger(__name__)
+
+
+class YahooGetter(models.Model):
+    _inherit = "currency.rate.provider"
+
+    @api.one
+    def _request_to_yahoo_finance(self):
+        YahooObj = self.search([('service_provider', '=', 'yahoo_finance')])
+        if YahooObj:
+            currencies_to_update = map(lambda x: x.name, self.currency_ids)
+            base_currency = map(lambda x: x.name, self.base_currency_id)
+
+            currency_pairs = ','.join(base_currency[0] + x for x in currencies_to_update)
+            request_url = "http://query.yahooapis.com/v1/public/yql?q=select * from yahoo.finance.xchange where pair in('{0}')&format=json&env=store://datatables.org/alltableswithkeys".format(currency_pairs)
+            parse_url = urllib.urlopen(request_url).read()
+            return parse_url
+
+    @api.one
+    def _parse_currency_from_yahoo_finance(self):
+        raw_data = json.dumps(self._request_to_yahoo_finance())
+        json_data = json.loads(raw_data)
+        try:
+            currency_vals = []
+            unlist_json = json.loads(json_data[0])
+            currencies_to_update = len(map(lambda x: x.name, self.currency_ids))
+            if currencies_to_update == 1:
+                currency = {}
+                currency['rate'] = unlist_json['query']['results']['rate']['Rate']
+                currency['currency_id'] = self.env['res.currency'].search([('name', 'ilike', unlist_json['query']['results']['rate']['Name'].split('/')[-1])]).id
+                currency_vals.append(currency)
+            elif currencies_to_update > 1:
+                for data in unlist_json['query']['results']['rate']:
+                    currency = {}
+                    currency['rate'] = data['Rate']
+                    currency['currency_id'] = self.env['res.currency'].search([('name', 'ilike', data['Name'].split('/')[-1])]).id
```

Comment:
`self.env['res.currency']` can be defined as variable rather than redefining everytime in loops and conditions.